rfcs/2021-11-xx-9862-ingest-apm-stats-along-traces-in-dd-agent-source.md

```diff
@@ -0,0 +1,164 @@
+
+Those stats are computed by a component named [concentrator] in the trace-agent. There is a [dedicated
+path][client-stats-aggregator] for APM stats that comes directly from tracing libraries. But ultimately they flow
+through datadog with the same [sending code][stats-writer].
+
+If we details a bit the actual content of a stats payload, it is a list of
+[`ClientGroupedStats`][client-grouped-stats-proto]:
+```
+	string service = 1;
+	string name = 2;
+	string resource = 3;
+	uint32 HTTP_status_code = 4;
+	string type = 5;
+	string DB_type = 6; // db_type might be used in the future to help in the obfuscation step
+	uint64 hits = 7; // count of all spans aggregated in the groupedstats
+	uint64 errors = 8; // count of error spans aggregated in the groupedstats
+	uint64 duration = 9; // total duration in nanoseconds of spans aggregated in the bucket
+	bytes okSummary = 10; // ddsketch summary of ok spans latencies encoded in protobuf
+	bytes errorSummary = 11; // ddsketch summary of error spans latencies encoded in protobuf
+	bool synthetics = 12; // set to true on spans generated by synthetics traffic
+	uint64 topLevelHits = 13; // count of top level spans aggregated in the groupedstats
+```
+
+So it is basically a group of various metrics that can be represented in Vector (Sketches are support since this
+[PR][sketch-pr]). In the proto definition sketches are stored as unstructured bytes slices, but those
+[fields][call-to-toproto] are filled with a [protobuf encoded ddsketch][ddsktech-proto]. Given the sketch implementation
+in Vector is also [heavily based on ddsketch][vector-sketch]  this should not cause any significant accuracy loss
+(**TBC**).
+
+This raises the question of having a second (assuming that the `datadog-agent` sources accepts Datadog Agent metrics -
+[RFC][dd-agent-metric-rfc]), unrelated, metric stream coming out of the `datadog-agent` source. Obviously metrics
+ingested from APM stats will have to be routed along with traces, and most often they will follow a different path that
+other plain metrics. Thus it is suggested to re-arrange the `datadog-agent` source according to the following points
+(Additional details on what exactly are "Datadog Agents" can be found in the trace support [RFC][trace-support-pr] and
+may provide relevant context for undermentionned points):
+
+* Keep one Vector source per kind of Datadog Agent
+* Then we would have the following Vector sources:
+  * `datadog-core`: would be the current `datadog-agent` sources, it would receive the data sent by the Datadog "core" Agent (the agent that collects logs & metrics). Note that renaming this source is probably optional, but documentation update would be required anyway.
+  * `datadog-trace`: would support all data sent by the `trace-agent`
+  * And so on as the support list grow:
+    * `datadog-process` source for the Datadog `process-agent`
+    * `datadog-security` for the Datadog security Agent
+    * Etc.
```

Comment:
This strikes me as a heavy approach that will lead to a lot of code duplication and boilerplate configuration for Vector users that want to send all data from their Datadog Agent to Vector.

Could we possibly consider an approach where we add the ability to store metadata in events as unstructured values?  We already want to be able to do that work, and this seems like a good use case for it.  This would, I believe, allow us to carry all of the type-special data in metadata fields, and sinks could be responsible for verifying that an event has the right schema (necessary base fields + necessary metadata fields) to be sent.

Reply:
I know this the whole opaque metadata idea has been around for a while, but I agree that emitting a LogEvent populated with most of the stats payload as fields and storing the two sketches undecoded as metadata would allow us to support most of the filtering/routing usecases without having to do extra conversion that would remain pretty much useless until further notice.

So bouncing on the metadata the naive approach would be to add an optional string/bytes map in metadata. I don't know if we have bigger planned in the near future plan but this would definitely be a good option here and much safer than pushing for arbitrary bytes sequence in a Value::Byte.